File: sdk/src/client/bindings/python/iota_client/types/unlock_condition.py
from enum import Enum


class UnlockConditionType(Enum):
    Address = 0
    StorageDepositReturn = 1
    Timelock = 2
    Expiration = 3
    StateControllerAddress = 4
    GovernorAddress = 5
    ImmutableAliasAddress = 6


class UnlockCondition():
    def __init__(self, type=None, address=None, amount=None, unix_time=None, return_address=None):
        """Initialize an UnlockCondition

        Parameters
        ----------
        type : UnlockConditionType
            The type of unlock condition
        address : Address
            Address for unlock condition
        amount : int
            Amount for storage deposit unlock condition
        unix_time : int
            Unix timestamp for timelock and expiration unlock condition
        return_address : Address
            Return address for expiration and storage deposit unlock condition
        """
        self.type = type
        self.address = address
        self.amount = amount
        self.unix_time = unix_time
        self.return_address = return_address

    def as_dict(self):
        config = {k: v for k, v in self.__dict__.items() if v != None}

        if 'type' in config:
            config['type'] = config['type'].value

        if 'address' in config:
            config['address'] = config['address'].as_dict()

        if 'return_address' in config:
            config['return_address'] = config['return_address'].as_dict()

        if 'amount' in config:
            config['amount'] = str(config['amount'])

        return config


class AddressUnlockCondition(UnlockCondition):
    def __init__(self, address):
        """Initialize an AddressUnlockCondition

        Parameters
        ----------
        address : Address
            Address
        """
        super().__init__(type=UnlockConditionType.Address, address=address)


class StorageDepositReturnUnlockCondition(UnlockCondition):
    def __init__(self, amount, return_address):
        """Initialize a StorageDepositReturnUnlockCondition

        Parameters
        ----------
        amount : int
            Amount
        return_address : Address
            Return address
        """
        super().__init__(type=UnlockConditionType.StorageDepositReturn,
                         amount=amount, return_address=return_address)


class TimelockUnlockCondition(UnlockCondition):
    def __init__(self, unix_time):
        """Initialize a TimelockUnlockCondition

        Parameters
        ----------
        unix_time : int
            Unix timestamp at which to unlock output
        """
        super().__init__(type=UnlockConditionType.Timelock, unix_time=unix_time)


class ExpirationUnlockCondition(UnlockCondition):
    def __init__(self, unix_time, return_address):
        """Initialize an ExpirationUnlockCondition

        Parameters
        ----------
        unix_time : int
            Unix timestamp
        return_address : Address
            Return address
        """
        super().__init__(type=UnlockConditionType.Expiration,
                         unix_time=unix_time, return_address=return_address)


class StateControllerAddressUnlockCondition(UnlockCondition):
    def __init__(self, address):
        """Initialize a StateControllerAddressUnlockCondition

        Parameters
        ----------
        address : Address
            Address for unlock condition
        """
        super().__init__(type=UnlockConditionType.StateControllerAddress, address=address)


class GovernorAddressUnlockCondition(UnlockCondition):
    def __init__(self, address):
        """Initialize a GovernorAddressUnlockCondition

        Parameters
        ----------
        address : Address
            Address for unlock condition
        """
        super().__init__(type=UnlockConditionType.GovernorAddress, address=address)


class ImmutableAliasAddressUnlockCondition(UnlockCondition):
    def __init__(self, address):
        """Initialize an ImmutableAliasAddressUnlockCondition

        Parameters
        ----------
        address : Address
            Address for unlock condition
        """
        super().__init__(type=UnlockConditionType.ImmutableAliasAddress, address=address)

File: bindings/nodejs/lib/types/unlock_condition.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { Address, AliasAddress } from './address';

/**
 * All of the unlock condition types.
 */
enum UnlockConditionType {
    Address = 0,
    StorageDepositReturn = 1,
    Timelock = 2,
    Expiration = 3,
    StateControllerAddress = 4,
    GovernorAddress = 5,
    ImmutableAliasAddress = 6,
}

abstract class UnlockCondition {
    private type: UnlockConditionType;

    constructor(type: UnlockConditionType) {
        this.type = type;
    }
    /**
     * The type of unlock condition.
     */
    getType(): UnlockConditionType {
        return this.type;
    }
}

class AddressUnlockCondition extends UnlockCondition /*implements IAddressUnlockCondition*/ {
    private address: Address;
    constructor(address: Address) {
        super(UnlockConditionType.Address);
        this.address = address;
    }

    /**
     * The address.
     */
    getAddress(): Address {
        return this.address;
    }
}
/**
 * Storage Desposit Return Unlock Condition.
 */
class StorageDepositReturnUnlockCondition extends UnlockCondition /*implements IStorageDepositReturnUnlockCondition*/ {
    private amount: string;
    private returnAddress: Address;

    constructor(returnAddress: Address, amount: string) {
        super(UnlockConditionType.StorageDepositReturn);
        this.amount = amount;
        this.returnAddress = returnAddress;
    }
    /**
     * Amount of tokens the consuming transaction must deposit to the address defined in return address.
     */
    getAmount(): string {
        return this.amount;
    }

    /**
     * The return address.
     */
    getReturnAddress(): Address {
        return this.returnAddress;
    }
}
/**
 * Timelock Unlock Condition.
 */
class TimelockUnlockCondition extends UnlockCondition /*implements ITimelockUnlockCondition*/ {
    private unixTime: number;

    constructor(unixTime: number) {
        super(UnlockConditionType.Timelock);
        this.unixTime = unixTime;
    }
    /**
     * Unix time (seconds since Unix epoch) starting from which the output can be consumed.
     */
    getUnixTime(): number {
        return this.unixTime;
    }
}

class ExpirationUnlockCondition extends UnlockCondition /*implements IExpirationUnlockCondition*/ {
    private returnAddress: Address;
    private unixTime: number;

    constructor(returnAddress: Address, unixTime: number) {
        super(UnlockConditionType.Expiration);
        this.returnAddress = returnAddress;
        this.unixTime = unixTime;
    }
    /**
     * Before this unix time, the condition is allowed to unlock the output,
     * after that only the address defined in return address.
     */
    getUnixTime(): number {
        return this.unixTime;
    }

    /**
     * The return address.
     */
    getReturnAddress(): Address {
        return this.returnAddress;
    }
}
/**
 * State Controller Address Unlock Condition.
 */
class StateControllerAddressUnlockCondition extends UnlockCondition /*implements IStateControllerAddressUnlockCondition*/ {
    private address: Address;
    constructor(address: Address) {
        super(UnlockConditionType.StateControllerAddress);
        this.address = address;
    }

    /**
     * The address.
     */
    getAddress(): Address {
        return this.address;
    }
}
/**
 * Governor Unlock Condition.
 */
class GovernorAddressUnlockCondition extends UnlockCondition /*implements IGovernorAddressUnlockCondition*/ {
    private address: Address;
    constructor(address: Address) {
        super(UnlockConditionType.GovernorAddress);
        this.address = address;
    }

    /**
     * The address.
     */
    getAddress(): Address {
        return this.address;
    }
}
/**
 * Immutable Alias Unlock Condition.
 */
class ImmutableAliasAddressUnlockCondition extends UnlockCondition /*implements IImmutableAliasAddressUnlockCondition*/ {
    private address: Address;
    constructor(address: AliasAddress) {
        super(UnlockConditionType.ImmutableAliasAddress);
        this.address = address;
    }

    /**
     * The address.
     */
    getAddress(): Address {
        return this.address;
    }
}

export {
    UnlockCondition,
    AddressUnlockCondition,
    StorageDepositReturnUnlockCondition,
    TimelockUnlockCondition,
    ExpirationUnlockCondition,
    StateControllerAddressUnlockCondition,
    GovernorAddressUnlockCondition,
    ImmutableAliasAddressUnlockCondition,
};
